### DISCUSSION TITLE
How are applications supposed to find libraries in /opt/local/lib and /opt/homebrew/lib ?

### DISCUSSION THREAD

**Discussion by larskanis** [Homebrew/discussions#1600] [Everyday usage] [answered]
People are using applications like ruby installed into their home directory, which make use of libraries installed by homebrew. Since homebrew doesn't install libraries into some system default library search path, but into `/opt/local/lib` and `/opt/homebrew/lib`, they aren't found without further measurements. That leaded to issues like https://github.com/libvips/ruby-vips/pull/144 in the past.

As a workaround Ruby-FFI implements a [hardcoded list](https://github.com/ffi/ffi/blob/82e20928319e499b2e60679205733cb5f088c363/lib/ffi/library.rb#L129) of search paths, which is used as a fallback, if otherwise library loading would fail. This list was extended to the above two directories in https://github.com/ffi/ffi/pull/882/files to enable loading of homebrew libraries.

Is there a better way to locate homebrew libraries other than hardcoding filesystem paths into depending applications/libraries?

Why are there no system library load mechanisms to make sure the homebrew libraries are found? On Linux there's `/etc/ld.so.conf.d/*.conf` which is used to extend paths used by the dynamic loader of the system.

Is the assumption correct at all, that homebrew libraries are expected to work when used by non-homebrew applications?

How do homebrew's own applications find depending libraries? Something like `LD_LIBRARY_PATH` on Linux?

I'm sorry, my knowledge about homebrew and MacOS is very limited.

**Comment by carlocab** [comment, marked as answer]
This is build-system dependent, unfortunately. Usually, setting `LDFLAGS=-L/opt/homebrew/lib` should work (as long as the library you're looking for is not from a keg-only formula).

**Reply by larskanis** [reply]
I have to clarify my question: FFI isn't build-system dependent, but [makes use](https://github.com/ffi/ffi/blob/82e20928319e499b2e60679205733cb5f088c363/ext/ffi_c/DynamicLibrary.c#L119) of `dl_open()` to load the library by the dynamic linker at runtime. Subsequently the functions are resolved by `dl_sym()`, but my questions above are about loading at runtime only.

**Reply by SMillerDev** [reply]
It just uses the system lookup functions: `LD_LIBRARY_PATH` will probably be all you need.

**Reply by carlocab** [reply]
I see. You can try using `install_name_tool` to change the path prefix of a library install name to `@rpath` then add the correct replacement for `@rpath` to the binary's `RPATH`s. (Both can be done with `install_name_tool`.)

If a linked library's install name is already prefixed by `@rpath`, there are also some environment variables you can set instead of hardcoding search paths into the binary. See `man dyld`. I don't recommend using this; this doesn't work the way you expect because of SIP.

Of course, using `install_name_tool` is also pretty much about hardcoding paths into the binary, so I guess the answer to your original question is no. (At least, not reliably.)

**Reply by larskanis** [reply]
I found that [some environment variables](https://github.com/Homebrew/brew/blob/d6efa5a76acd05357a101704258086d63e4e7b55/Library/Homebrew/cmd/shellenv.sh) are set by homebrew to enable homebrew binaries. Unfortunately there's nothing like `LD_LIBRARY_PATH` that points to the `${HOMEBREW_PREFIX}/lib` directory. And even if there would be such a variable it would be disregarded by system executables (like ruby) as described [here](https://github.com/libvips/ruby-vips/pull/144#issuecomment-395965751).

How about looking up into `${HOMEBREW_PREFIX}/lib` by ruby-ffi (instead of absolute path)? Can we expect to have the `HOMEBREW_PREFIX` environment variable defined by all users of homebrew? Or is this optional?

**Reply by carlocab** [reply]
This is optional. You can get this with `brew --prefix`. You’ll want to cache this value if you use it in several places.

**Reply by larskanis** [reply]
Thank you! I think this answers my question.

**Reply by carlocab** [reply]
Sure. As an aside, adding `RPATH` entries to a binary object that calls `dlopen` (no `_`) is still useful. For example, I can `dlopen("libcurl.dylib")` and find Homebrew's `libcurl` if I've added `$(brew --prefix)/opt/curl/lib` to my binary's `RPATH`s (as long as I haven't added `/usr/lib` as an `RPATH` as well -- but that's in the default library search path anyway).

Not really familiar with the details of `dl_open`, but it's probably still just `dlopen` underneath.

**Reply by larskanis** [reply]
Sorry, I meant the well known `dlopen()` call - the `dl_open()` with underscore is just a ruby-ffi internal alias.
